lib/pure/asyncmacro.nim

```diff
@@ -126,9 +126,16 @@ template await*(f: typed): untyped {.used.} =
     error "await expects Future[T], got " & $typeof(f)
 
 template await*[T](f: Future[T]): auto {.used.} =
-  var internalTmpFuture: FutureBase = f
-  yield internalTmpFuture
-  (cast[typeof(f)](internalTmpFuture)).read()
+  template yieldFuture = yield FutureBase()
+
+  when compiles(yieldFuture):
+    var internalTmpFuture: FutureBase = f
+    yield internalTmpFuture
+    (cast[typeof(f)](internalTmpFuture)).read()
+  else:
+    static: error(
+      "Can only 'await' inside a proc marked as 'async'. Use " &
+      "'waitFor' when calling an 'async' proc in a non-async scope instead")
```

Comment:
Try something like below to get better error message report
```nim
macro errorAsync[T](yieldFuture: Future[T]) =
  error(
        "Can only 'await' inside a proc marked as 'async'. Use " &
        "'waitFor' when calling an 'async' proc in a non-async scope instead",
        yieldFuture)

template await*[T](f: Future[T]): auto {.used.} =
  template yieldFuture = yield FutureBase()

  when compiles(yieldFuture):
    var internalTmpFuture: FutureBase = f
    yield internalTmpFuture
    (cast[typeof(f)](internalTmpFuture)).read()
  else:
    errorAsync(f)
```

> test.nim(68, 8) Error: Can only 'await' inside a proc marked as 'async'. Use 'waitFor' when calling an 'async' proc in a non-async scope instead

Reply:
Oh I forgot about that, that's much nicer! if only there was a way to get the NimNode for the entire template so IDEs like VSCode would highlight the entire `await` statement.